### PR TITLE
PR: Inherit from JavaScript in Jupyter Notebook v7

### LINKED ISSUE #330: Upgrade Jupyter Lab to version 3
Incorporate all changes to the Notebook example in Jupyter Lab version 2 in here where relevant.

This is also the proper fix for #319.

### PR DESCRIPTION
**Pull request by jitseniesen**
This is a further iteration of #395. Instead of using the Jupyter Notebook v7 server as is, without any JavaScript of our own, we write our own JavaScript inheriting from the JavaScript classes in Jupyter Notebook v7. This makes our code more complicated and also more tightly coupled to Jupyter Notebook, but it gives us a lot more flexibility to do what we want.

Fixes #330.

### CODE REVIEW

**Comment by jitseniesen** [comment]
The code and the commit history is in a mess right now, but I feel I am making progress and I am hopeful that the general framework is now right. In particular, I can hide parts of the notebook (the top panel and the File menu) fairly easily, as shown in the screenshot below.

notebook-inherit

**Comment by jitseniesen** [comment]
All the basic stuff is working now as it was before. The following animation shows how to create a new notebook, write some commands in it, attach a console, show the variables in Spyder's Variable Explorer, and save the notebook in another directory.

spynb-run

As I see it, there are two big advantages to basing our plugin on Jupyter Notebook. By leveraging their code, there is less code for us to write, so it will be easier to maintain our plugin. Concretely, this means that we can use Jupyter Lab extensions. This PR is based on Jupyter Lab v4 which is in alpha at the moment, so many extensions have not made the transition. I tried plotly, but that does not seem to work properly. Then I tried ipywidgets, which gives many errors about version mismatch but it does seem to work.

spynb-lor

As far as I am concerned, this PR is getting close to where it can be merged. I tidied up the code and the commit history. Here are some remaining issues:
* The Windows tests fail on CI. I haven't yet tried the PR on Windows, but there is a good chance that it does not work there. This needs to be checked before merging.
* More generally, the tests on CI are temperamental and fail regularly. This has always been the case.
* In the current version of the plugin, right clicking brings up Spyder's context menu (with New, Open, Save); in the PR, something has changed and it brings up Jupyter's context menu. I am not yet sure what to do about this: all actions in Spyder's context menu can also be performed via the hamburger menu and Jupyter's context menu does have some useful stuff in it.
* The current version of a plugin has an option for choosing light or dark mode (in the Spyder plugin preferences). The implementation was very hacky and no longer works. Users can now switch from light to dark using the Jupyter menu in the notebook. I think we can just remove the Spyder plugin preference for light or dark.
* It may be possible to separate the Jupyter and Spyder environments. This needs discussion about whether that's a good idea.
* Some of the options in the Jupyter notebook menus probably don't work inside Spyder (e.g., switch to Jupyter Lab) or are not relevant. 

My current thinking is to resolve the first bullet point and then merge the PR, opening issues for the other bullet point to leave them to later.

**Comment by jitseniesen** [comment]
The plugin does seem to work on Windows when I test manually, so it is "only" a matter of getting the tests to pass.

File: spyder_notebook/server/notebook-template.html [hunk 1]
```diff
@@ -0,0 +1,39 @@
+<!DOCTYPE html>
```
Comment (ccordoba12) [inline spyder_notebook/server/notebook-template.html:1]: Could you add in a comment above this line why this template is necessary? That's not clear from the content below.

File: spyder_notebook/utils/servermanager.py [hunk 2]
```diff
@@ -190,15 +191,12 @@ def start_server(self, filename, interpreter):
...
-        arguments = [serverscript, '--no-browser',
+        arguments = ['-m', 'spyder_notebook.server', '--no-browser',
                      '--notebook-dir={}'.format(nbdir),
-                     '--NotebookApp.password=',
+                     '--ServerApp.password=',
                      '--KernelSpecManager.kernel_spec_class={}'.format(
                            KERNELSPEC)]
-        if self.dark_theme:
-            arguments.append('--dark')
+        # TODO: Add support for dark theme
```
Comment (ccordoba12) [inline spyder_notebook/utils/servermanager.py:199]:
```suggestion

        # TODO: Add support for dark theme
```

File: spyder_notebook/utils/servermanager.py [hunk 3]
```diff
@@ -279,7 +278,15 @@ def shutdown_all_servers(self):
                              server.notebook_dir)
                 server.process.errorOccurred.disconnect()
                 server.process.finished.disconnect()
-                notebookapp.shutdown_server(server.server_info)
+                try:
```
Comment (ccordoba12) [inline spyder_notebook/utils/servermanager.py:281]:
```suggestion

                try:
```

File: spyder_notebook/utils/tests/test_servermanager.py [hunk 4]
```diff
@@ -74,12 +74,12 @@ def test_get_server_with_server(
...
-                         [(True, True), (False, True), (False, False)])
-def test_start_server(mocker, dark, under_home):
-    """Test that .start_server() starts a process with the correct script file,
-    notebook dir, and dark argument; that it stores the server process in
-    `.servers`; and that it calls ._check_server_running()."""
+@pytest.mark.parametrize('under_home', [True, False])
+def test_start_server(mocker, under_home):
+    """Test that .start_server() starts a process with the correct arguments,
+    that it stores the server process in `.servers`; and that it calls
+    ._check_server_running()."""
```
Comment (ccordoba12) [inline spyder_notebook/utils/tests/test_servermanager.py:81]:
```suggestion
    """
    Test that .start_server() starts a process with the correct arguments,
    that it stores the server process in `.servers`; and that it calls
    ._check_server_running().
    """
```

File: spyder_notebook/widgets/notebooktabwidget.py [hunk 5]
```diff
@@ -51,6 +52,38 @@
...
+    the file is in use by another process, this function waits half a
+    second in that case and tries again. After three attempts, it gives up.
+
+    Parameters
+    ----------
+    filename : str
+        Name of file which should be deleted
+    """
+    if sys.platform != 'win32':
+            os.remove(filename)
```
Comment (ccordoba12) [inline spyder_notebook/widgets/notebooktabwidget.py:74]:
```suggestion
    if sys.platform != 'win32':
        os.remove(filename)
```

File: spyder_notebook/widgets/tests/test_client.py [hunk 6]
```diff
@@ -41,13 +41,14 @@ def plugin(plugin_without_server):
...
     """
-    server_info = {'notebook_dir': '/path/notebooks',
+    server_info = {'root_dir': '/path/notebooks',
                    'url': 'fake_url',
                    'token': 'fake_token'}
     plugin_without_server.client.register(server_info)
     return plugin_without_server
 
 
+@pytest.mark.skip
```
Comment (ccordoba12) [inline spyder_notebook/widgets/tests/test_client.py:51]: Do you plan to address this later?

**Review by ccordoba12 — COMMENTED** [review]
Thanks @jitseniesen for all your hard work on this! It looks really really cool!!

File: spyder_notebook/widgets/tests/test_client.py (hunk 6 above)
Comment (jitseniesen) [reply spyder_notebook/widgets/tests/test_client.py:51]: Well spotted! I don't think there is a reason anymore to skip that test, so I'll enable it again.

**Comment by jitseniesen** [comment]
I rebased so that PR #408 (bug fix for venv under Windows) is incorporated. All review comments are addressed. I think this is ready to go in.

**Pull request merged**